source/common/common/phantom.h

```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <utility>
+
+namespace Envoy {
+
+// A phantom type allows wrapping a common type with additional type information in order to allow
+// additional compile time safety when passing it around.
+template <class InnerT, class TagT> struct Phantom {
+  Phantom() = default;
+  explicit Phantom(const InnerT& t) : val_(t) {}
+  explicit Phantom(InnerT&& t) : val_(std::move(t)) {}
+
+  InnerT& get() { return val_; }
```

Comment:
As I mentioned previously, if you want it to be even more succinct, you could consider overloading `operator()` instead of `get()`. Personally, I prefer this since it's less magical, but just throwing it out there.

Reply:
Yeah I agree on making this less magical, `operator()` makes it look like a function object